### DISCUSSION TITLE
Why MMAP_SYSCALL can be used inside a signal handler?Is there any problem？

### DISCUSSION THREAD

**Discussion by hustchaya** [async-profiler/async-profiler#915] [General]
void* OS::safeAlloc(size_t size) {
    // Naked syscall can be used inside a signal handler.
    // Also, we don't want to catch our own calls when profiling mmap.
    intptr_t result = syscall(MMAP_SYSCALL, NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (result < 0 && result > -4096) {
        return NULL;
    }
    return (void*)result;
}

**Comment by apangin** [comment]
A syscall, unlike a libray call, is atomic from the application perspective.
